src/gsad.c

```diff
@@ -3972,19 +4091,25 @@ send_redirect_to_urn (struct MHD_Connection *connection, const char *urn,
 
   host = MHD_lookup_connection_value (connection, MHD_HEADER_KIND,
                                       MHD_HTTP_HEADER_HOST);
-  if (host && g_utf8_validate (host, -1, NULL) == FALSE)
-    {
-      send_response (connection,
-                     UTF8_ERROR_PAGE ("'Host' header"),
-                     MHD_HTTP_BAD_REQUEST, NULL,
-                     GSAD_CONTENT_TYPE_TEXT_HTML, NULL, 0);
-      return MHD_YES;
-    }
-  if (host == NULL)
+
+  switch (validate_host_header (host))
     {
-      send_response (connection, BAD_REQUEST_PAGE, MHD_HTTP_NOT_ACCEPTABLE,
-                     NULL, GSAD_CONTENT_TYPE_TEXT_HTML, NULL, 0);
-      return MHD_YES;
+      case 0:
+        // header is valid
+        break;
+      case 1:
+        // Invalid UTF-8
+        send_response (connection,
+                       UTF8_ERROR_PAGE ("'Host' header"),
+                       MHD_HTTP_BAD_REQUEST, NULL,
+                       GSAD_CONTENT_TYPE_TEXT_HTML, NULL, 0);
+        return MHD_YES;
+      case 2:
+      default:
+        // Otherwise invalid
+        send_response (connection, BAD_REQUEST_PAGE, MHD_HTTP_NOT_ACCEPTABLE,
```

Comment:
For improved UX i wouldn't throw a bad request around if the user just accessed the system via a wrong Hostname/IP. If the accepted host headers are just misconfigured the user will never now directly why he gets a bad request when accessing the system.

A way better approach would be to throw a 403 / Forbidden and include a short note in the returned HTML document explaining how to configure the system in a way to be able to access it.